binrw/src/binread/impls.rs

```diff
@@ -29,12 +31,91 @@ macro_rules! binread_impl {
                         }
                     })
                 }
+
+                $(
+                    $($is_int)?
+                    fn read_options_count<'a, R, C>(
+                        reader: &mut R,
+                        endian: Endian,
+                        _args: Self::Args<'a>,
+                        count: C::Count,
+                    ) -> BinResult<C>
+                    where
+                        R: Read + Seek,
+                        Self::Args<'a>: Clone,
+                        C: crate::container::Container<Item=Self>,
+                    {
+                        C::new_smart(count, |mut buf| {
+                            reader.read_exact(&mut bytemuck::cast_slice_mut::<_, u8>(
+                                &mut buf,
+                            ))?;
+
+                            if core::mem::size_of::<Self>() != 1
+                                && ((cfg!(target_endian = "big") && endian == crate::Endian::Little)
+                                    || (cfg!(target_endian = "little") && endian == crate::Endian::Big))
+                            {
+                                for value in buf.iter_mut() {
+                                    *value = value.swap_bytes();
+                                }
+                            }
+                            Ok(())
+                        })
+                    }
+                )?
             }
         )*
     }
 }
 
-binread_impl!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, f32, f64);
+binread_impl!(
+    u16 as int,
+    u32 as int,
+    u64 as int,
+    u128 as int,
+    i8 as int,
+    i16 as int,
+    i32 as int,
+    i64 as int,
+    i128 as int,
+    f32,
+    f64
+);
+
+impl BinRead for u8 {
+    type Args<'a> = ();
+    fn read_options<R: Read + Seek>(
+        reader: &mut R,
+        _endian: Endian,
+        _args: Self::Args<'_>,
+    ) -> BinResult<Self> {
+        let mut val = 0u8;
+        let pos = reader.stream_position()?;
+        reader
+            .read_exact(core::slice::from_mut(&mut val))
+            .or_else(crate::__private::restore_position(reader, pos))?;
+
+        Ok(val)
+    }
+
+    // This extra impl for `u8` makes it faster than
+    // `binread_impl`, but *only* because `binread_impl` is not allowed
+    // to use unsafe code to eliminate the unnecessary zero-fill.
+    // Otherwise, performance would be identical and it could be
+    // deleted.
+    fn read_options_count<'a, R, C>(
+        reader: &mut R,
+        _endian: Endian,
+        _args: Self::Args<'a>,
+        count: C::Count,
+    ) -> BinResult<C>
+    where
+        R: Read + Seek,
+        Self::Args<'a>: Clone,
+        C: crate::container::Container<Item = Self>,
+    {
+        C::new_smart(count, |buf| reader.read_exact(buf).map_err(Into::into))
+    }
```

Comment:
Oops. This comment was copied over and is inaccurate now because I wasn't paying close enough attention to how I was implementing this version of the specialization. (Notably, we can only get this version of the specialization when using Vec, and I've completely abstracted access to that Vec away here. This is a performance regression.

Ideally we'd be using the current implementation for arrays, and read_to_end for Vec. Funnily enough, we might be able to use fake specialization that requires 'static here to switch between the two implementations, since we only need to care about u8 (and anything that might implement `Container`... should we be sealing `Container`?) instead of <all possible user created types>.

Reply:
Actually, no, that doesn't work. That would require us to bound `C: Container<Item=Self> + 'static`, and if user generated items are not static, the `Container` wrapping those items probably won't be static either.